File: sol-trade-sdk-nodejs/src/instruction/raydium_cpmm_builder.ts
/**
 * Raydium CPMM (Concentrated Pool Market Maker) Protocol Instruction Builder
 *
 * Production-grade instruction builder for Raydium CPMM protocol.
 * Supports swap operations with WSOL and USDC pools.
 */

import {
  PublicKey,
  Keypair,
  AccountMeta,
  Instruction,
  SystemProgram,
} from "@solana/web3.js";
import {
  getAssociatedTokenAddressSync,
  createAssociatedTokenAccountInstruction,
  TOKEN_PROGRAM_ID,
  createCloseAccountInstruction,
  NATIVE_MINT,
  createSyncNativeInstruction,
} from "@solana/spl-token";

// ============================================
// Program IDs and Constants
// ============================================

/** Raydium CPMM program ID */
export const RAYDIUM_CPMM_PROGRAM_ID = new PublicKey(
  "CPMMoo8L3F4NbTegBCKVNunggL7H1ZpdTHKxQB5qKP1C"
);

/** Authority */
export const AUTHORITY = new PublicKey(
  "GpMZbSM2GgvTKHJirzeGfMFoaZ8UR2X7F4v8vHTvxFbL"
);

/** WSOL Token Account (mint) */
export const WSOL_TOKEN_ACCOUNT = new PublicKey(
  "So11111111111111111111111111111111111111112"
);

/** USDC Token Account (mint) */
export const USDC_TOKEN_ACCOUNT = new PublicKey(
  "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
);

/** Fee rates */
export const FEE_RATE_DENOMINATOR_VALUE = 1_000_000n;
export const TRADE_FEE_RATE = 2500n;
export const CREATOR_FEE_RATE = 0n;
export const PROTOCOL_FEE_RATE = 120000n;
export const FUND_FEE_RATE = 40000n;

// ============================================
// Discriminators
// ============================================

/** Swap base in instruction discriminator */
export const SWAP_BASE_IN_DISCRIMINATOR: Buffer = Buffer.from([
  143, 190, 90, 218, 196, 30, 51, 222,
]);

/** Swap base out instruction discriminator */
export const SWAP_BASE_OUT_DISCRIMINATOR: Buffer = Buffer.from([
  55, 217, 98, 86, 163, 74, 180, 173,
]);

// ============================================
// Seeds
// ============================================

export const POOL_SEED = Buffer.from("pool");
export const POOL_VAULT_SEED = Buffer.from("pool_vault");
export const OBSERVATION_STATE_SEED = Buffer.from("observation");

// ============================================
// PDA Derivation Functions
// ============================================

/**
 * Derive the pool PDA for given config and mints
 */
export function getPoolPda(
  ammConfig: PublicKey,
  mint1: PublicKey,
  mint2: PublicKey
): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [POOL_SEED, ammConfig.toBuffer(), mint1.toBuffer(), mint2.toBuffer()],
    RAYDIUM_CPMM_PROGRAM_ID
  );
  return pda;
}

/**
 * Derive the vault PDA for a pool and mint
 */
export function getVaultPda(poolState: PublicKey, mint: PublicKey): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [POOL_VAULT_SEED, poolState.toBuffer(), mint.toBuffer()],
    RAYDIUM_CPMM_PROGRAM_ID
  );
  return pda;
}

/**
 * Derive the observation state PDA for a pool
 */
export function getObservationStatePda(poolState: PublicKey): PublicKey {
  const [pda] = PublicKey.findProgramAddressSync(
    [OBSERVATION_STATE_SEED, poolState.toBuffer()],
    RAYDIUM_CPMM_PROGRAM_ID
  );
  return pda;
}

// ============================================
// Helper Functions
// ============================================

/**
 * Compute swap amount for CPMM
 */
export function computeSwapAmount(
  baseReserve: bigint,
  quoteReserve: bigint,
  isBaseIn: boolean,
  amountIn: bigint,
  slippageBasisPoints: bigint
): { amountOut: bigint; minAmountOut: bigint } {
  // Apply trade fee (0.25%)
  const feeRate = TRADE_FEE_RATE;
  const feeDenominator = FEE_RATE_DENOMINATOR_VALUE;
  const amountInAfterFee = amountIn - (amountIn * feeRate) / feeDenominator;

  // Calculate output using constant product formula
  let amountOut: bigint;
  if (isBaseIn) {
    // Selling base for quote: output = (quoteReserve * amountIn) / (baseReserve + amountIn)
    const denominator = baseReserve + amountInAfterFee;
    amountOut = (quoteReserve * amountInAfterFee) / denominator;
  } else {
    // Selling quote for base: output = (baseReserve * amountIn) / (quoteReserve + amountIn)
    const denominator = quoteReserve + amountInAfterFee;
    amountOut = (baseReserve * amountInAfterFee) / denominator;
  }

  // Apply slippage
  const minAmountOut = amountOut - (amountOut * slippageBasisPoints) / 10000n;

  return { amountOut, minAmountOut };
}

/**
 * Create instructions to wrap SOL into WSOL
 */
export function createWsolInstructions(
  payer: PublicKey,
  amount: bigint
): Instruction[] {
  const instructions: Instruction[] = [];
  const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payer, true);

  // Sync native (wrap SOL)
  instructions.push(createSyncNativeInstruction(wsolAta));

  return instructions;
}

/**
 * Create instruction to close WSOL ATA and unwrap to SOL
 */
export function createCloseWsolInstruction(payer: PublicKey): Instruction {
  const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payer, true);
  return createCloseAccountInstruction(wsolAta, payer, payer, [], TOKEN_PROGRAM_ID);
}

/**
 * Create WSOL ATA instruction
 */
export function createWsolAtaInstruction(payer: PublicKey): Instruction {
  const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payer, true);
  return createAssociatedTokenAccountInstruction(
    payer,
    wsolAta,
    payer,
    NATIVE_MINT,
    TOKEN_PROGRAM_ID
  );
}

// ============================================
// Types
// ============================================

export interface RaydiumCpmmParams {
  poolState?: PublicKey;
  ammConfig: PublicKey;
  baseMint: PublicKey;
  quoteMint: PublicKey;
  baseTokenProgram: PublicKey;
  quoteTokenProgram: PublicKey;
  baseVault?: PublicKey;
  quoteVault?: PublicKey;
  baseReserve: bigint;
  quoteReserve: bigint;
  observationState?: PublicKey;
}

export interface BuildBuyInstructionsParams {
  payer: Keypair | PublicKey;
  outputMint: PublicKey;
  inputAmount: bigint;
  slippageBasisPoints?: bigint;
  fixedOutputAmount?: bigint;
  createInputMintAta?: boolean;
  createOutputMintAta?: boolean;
  closeInputMintAta?: boolean;
  protocolParams: RaydiumCpmmParams;
}

export interface BuildSellInstructionsParams {
  payer: Keypair | PublicKey;
  inputMint: PublicKey;
  inputAmount: bigint;
  slippageBasisPoints?: bigint;
  fixedOutputAmount?: bigint;
  createOutputMintAta?: boolean;
  closeOutputMintAta?: boolean;
  closeInputMintAta?: boolean;
  protocolParams: RaydiumCpmmParams;
}

// ============================================
// Instruction Builders
// ============================================

/**
 * Build buy instructions for Raydium CPMM protocol
 */
export function buildBuyInstructions(
  params: BuildBuyInstructionsParams
): Instruction[] {
  const {
    payer,
    outputMint,
    inputAmount,
    slippageBasisPoints = 1000n,
    fixedOutputAmount,
    createInputMintAta = true,
    createOutputMintAta = true,
    closeInputMintAta = false,
    protocolParams,
  } = params;

  if (inputAmount === 0n) {
    throw new Error("Amount cannot be zero");
  }

  const payerPubkey = payer instanceof Keypair ? payer.publicKey : payer;
  const instructions: Instruction[] = [];

  const {
    ammConfig,
    baseMint,
    quoteMint,
    baseTokenProgram,
    quoteTokenProgram,
    baseVault,
    quoteVault,
    baseReserve,
    quoteReserve,
    observationState,
  } = protocolParams;

  // Check pool type
  const isWsol = baseMint.equals(WSOL_TOKEN_ACCOUNT) || quoteMint.equals(WSOL_TOKEN_ACCOUNT);
  const isUsdc = baseMint.equals(USDC_TOKEN_ACCOUNT) || quoteMint.equals(USDC_TOKEN_ACCOUNT);

  if (!isWsol && !isUsdc) {
    throw new Error("Pool must contain WSOL or USDC");
  }

  // Determine swap direction
  const isBaseIn = baseMint.equals(WSOL_TOKEN_ACCOUNT) || baseMint.equals(USDC_TOKEN_ACCOUNT);
  const mintTokenProgram = isBaseIn ? quoteTokenProgram : baseTokenProgram;

  // Derive pool state
  const poolState = protocolParams.poolState && !protocolParams.poolState.equals(PublicKey.default)
    ? protocolParams.poolState
    : getPoolPda(ammConfig, baseMint, quoteMint);

  // Calculate output
  const swapResult = computeSwapAmount(
    baseReserve,
    quoteReserve,
    isBaseIn,
    inputAmount,
    slippageBasisPoints
  );
  const minimumAmountOut = fixedOutputAmount || swapResult.minAmountOut;

  // Determine input/output mints
  const inputMint = isWsol ? WSOL_TOKEN_ACCOUNT : USDC_TOKEN_ACCOUNT;

  // Derive user token accounts
  const inputTokenAccount = getAssociatedTokenAddressSync(
    inputMint,
    payerPubkey,
    true,
    TOKEN_PROGRAM_ID
  );
  const outputTokenAccount = getAssociatedTokenAddressSync(
    outputMint,
    payerPubkey,
    true,
    mintTokenProgram
  );

  // Derive vault accounts
  const inputVaultAccount = getVaultPda(poolState, inputMint);
  const outputVaultAccount = getVaultPda(poolState, outputMint);

  // Derive observation state
  const observationStateAccount = observationState && !observationState.equals(PublicKey.default)
    ? observationState
    : getObservationStatePda(poolState);

  // Handle WSOL wrapping
  if (createInputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createAssociatedTokenAccountInstruction(
        payerPubkey,
        wsolAta,
        payerPubkey,
        NATIVE_MINT,
        TOKEN_PROGRAM_ID
      )
    );
    instructions.push(createSyncNativeInstruction(wsolAta));
  }

  // Create output mint ATA if needed
  if (createOutputMintAta) {
    instructions.push(
      createAssociatedTokenAccountInstruction(
        payerPubkey,
        outputTokenAccount,
        payerPubkey,
        outputMint,
        mintTokenProgram
      )
    );
  }

  // Build instruction data
  const data = Buffer.alloc(24);
  SWAP_BASE_IN_DISCRIMINATOR.copy(data, 0);
  data.writeBigUInt64LE(inputAmount, 8);
  data.writeBigUInt64LE(minimumAmountOut, 16);

  // Build accounts
  const accounts: AccountMeta[] = [
    { pubkey: payerPubkey, isSigner: true, isWritable: true },
    { pubkey: AUTHORITY, isSigner: false, isWritable: false },
    { pubkey: ammConfig, isSigner: false, isWritable: false },
    { pubkey: poolState, isSigner: false, isWritable: true },
    { pubkey: inputTokenAccount, isSigner: false, isWritable: true },
    { pubkey: outputTokenAccount, isSigner: false, isWritable: true },
    { pubkey: inputVaultAccount, isSigner: false, isWritable: true },
    { pubkey: outputVaultAccount, isSigner: false, isWritable: true },
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    { pubkey: mintTokenProgram, isSigner: false, isWritable: false },
    { pubkey: inputMint, isSigner: false, isWritable: false },
    { pubkey: outputMint, isSigner: false, isWritable: false },
    { pubkey: observationStateAccount, isSigner: false, isWritable: true },
  ];

  instructions.push(
    new Instruction({
      keys: accounts,
      programId: RAYDIUM_CPMM_PROGRAM_ID,
      data,
    })
  );

  // Close WSOL ATA if requested
  if (closeInputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createCloseAccountInstruction(wsolAta, payerPubkey, payerPubkey, [], TOKEN_PROGRAM_ID)
    );
  }

  return instructions;
}

/**
 * Build sell instructions for Raydium CPMM protocol
 */
export function buildSellInstructions(
  params: BuildSellInstructionsParams
): Instruction[] {
  const {
    payer,
    inputMint,
    inputAmount,
    slippageBasisPoints = 1000n,
    fixedOutputAmount,
    createOutputMintAta = true,
    closeOutputMintAta = false,
    closeInputMintAta = false,
    protocolParams,
  } = params;

  if (inputAmount === 0n) {
    throw new Error("Amount cannot be zero");
  }

  const payerPubkey = payer instanceof Keypair ? payer.publicKey : payer;
  const instructions: Instruction[] = [];

  const {
    ammConfig,
    baseMint,
    quoteMint,
    baseTokenProgram,
    quoteTokenProgram,
    baseReserve,
    quoteReserve,
    observationState,
  } = protocolParams;

  // Check pool type
  const isWsol = baseMint.equals(WSOL_TOKEN_ACCOUNT) || quoteMint.equals(WSOL_TOKEN_ACCOUNT);
  const isUsdc = baseMint.equals(USDC_TOKEN_ACCOUNT) || quoteMint.equals(USDC_TOKEN_ACCOUNT);

  if (!isWsol && !isUsdc) {
    throw new Error("Pool must contain WSOL or USDC");
  }

  // Determine swap direction
  const isQuoteOut = quoteMint.equals(WSOL_TOKEN_ACCOUNT) || quoteMint.equals(USDC_TOKEN_ACCOUNT);
  const mintTokenProgram = isQuoteOut ? baseTokenProgram : quoteTokenProgram;

  // Derive pool state
  const poolState = protocolParams.poolState && !protocolParams.poolState.equals(PublicKey.default)
    ? protocolParams.poolState
    : getPoolPda(ammConfig, baseMint, quoteMint);

  // Calculate output
  const swapResult = computeSwapAmount(
    baseReserve,
    quoteReserve,
    isQuoteOut,
    inputAmount,
    slippageBasisPoints
  );
  const minimumAmountOut = fixedOutputAmount || swapResult.minAmountOut;

  // Determine output mint
  const outputMint = isWsol ? WSOL_TOKEN_ACCOUNT : USDC_TOKEN_ACCOUNT;

  // Derive user token accounts
  const inputTokenAccount = getAssociatedTokenAddressSync(
    inputMint,
    payerPubkey,
    true,
    mintTokenProgram
  );
  const outputTokenAccount = getAssociatedTokenAddressSync(
    outputMint,
    payerPubkey,
    true,
    TOKEN_PROGRAM_ID
  );

  // Derive vault accounts
  const inputVaultAccount = getVaultPda(poolState, inputMint);
  const outputVaultAccount = getVaultPda(poolState, outputMint);

  // Derive observation state
  const observationStateAccount = observationState && !observationState.equals(PublicKey.default)
    ? observationState
    : getObservationStatePda(poolState);

  // Create WSOL ATA for receiving if needed
  if (createOutputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createAssociatedTokenAccountInstruction(
        payerPubkey,
        wsolAta,
        payerPubkey,
        NATIVE_MINT,
        TOKEN_PROGRAM_ID
      )
    );
  }

  // Build instruction data
  const data = Buffer.alloc(24);
  SWAP_BASE_IN_DISCRIMINATOR.copy(data, 0);
  data.writeBigUInt64LE(inputAmount, 8);
  data.writeBigUInt64LE(minimumAmountOut, 16);

  // Build accounts
  const accounts: AccountMeta[] = [
    { pubkey: payerPubkey, isSigner: true, isWritable: true },
    { pubkey: AUTHORITY, isSigner: false, isWritable: false },
    { pubkey: ammConfig, isSigner: false, isWritable: false },
    { pubkey: poolState, isSigner: false, isWritable: true },
    { pubkey: inputTokenAccount, isSigner: false, isWritable: true },
    { pubkey: outputTokenAccount, isSigner: false, isWritable: true },
    { pubkey: inputVaultAccount, isSigner: false, isWritable: true },
    { pubkey: outputVaultAccount, isSigner: false, isWritable: true },
    { pubkey: mintTokenProgram, isSigner: false, isWritable: false },
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    { pubkey: inputMint, isSigner: false, isWritable: false },
    { pubkey: outputMint, isSigner: false, isWritable: false },
    { pubkey: observationStateAccount, isSigner: false, isWritable: true },
  ];

  instructions.push(
    new Instruction({
      keys: accounts,
      programId: RAYDIUM_CPMM_PROGRAM_ID,
      data,
    })
  );

  // Close WSOL ATA if requested
  if (closeOutputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createCloseAccountInstruction(wsolAta, payerPubkey, payerPubkey, [], TOKEN_PROGRAM_ID)
    );
  }

  // Close input token ATA if requested
  if (closeInputMintAta) {
    instructions.push(
      createCloseAccountInstruction(
        inputTokenAccount,
        payerPubkey,
        payerPubkey,
        [],
        mintTokenProgram
      )
    );
  }

  return instructions;
}

File: sol-trade-sdk-python/src/instruction/raydium_cpmm_builder.py
"""
Raydium CPMM instruction builder for Solana trading SDK.
Production-grade implementation with all constants, discriminators, and PDA derivation functions.
"""

from typing import List, Optional
from dataclasses import dataclass
from solders.pubkey import Pubkey
from solders.instruction import Instruction, AccountMeta
import struct

from .common import (
    SYSTEM_PROGRAM,
    TOKEN_PROGRAM,
    TOKEN_PROGRAM_2022,
    WSOL_TOKEN_ACCOUNT,
    USDC_TOKEN_ACCOUNT,
    DEFAULT_SLIPPAGE,
    get_associated_token_address,
    create_associated_token_account_idempotent_instruction,
    handle_wsol,
    close_wsol,
    close_token_account_instruction,
    calculate_with_slippage_sell,
)

# ============================================
# Raydium CPMM Program ID
# ============================================

RAYDIUM_CPMM_PROGRAM_ID: Pubkey = Pubkey.from_string("CPMMoo8L3F4NbTegBCKVNunggL7H1ZpdTHKxQB5qKP1C")

# ============================================
# Raydium CPMM Constants
# ============================================

# Authority
AUTHORITY: Pubkey = Pubkey.from_string("GpMZbSM2GgvTKHJirzeGfMFoaZ8UR2X7F4v8vHTvxFbL")

# Fee Rates
FEE_RATE_DENOMINATOR_VALUE: int = 1_000_000
TRADE_FEE_RATE: int = 2500
CREATOR_FEE_RATE: int = 0
PROTOCOL_FEE_RATE: int = 120000
FUND_FEE_RATE: int = 40000

# ============================================
# Instruction Discriminators
# ============================================

SWAP_BASE_IN_DISCRIMINATOR: bytes = bytes([143, 190, 90, 218, 196, 30, 51, 222])
SWAP_BASE_OUT_DISCRIMINATOR: bytes = bytes([55, 217, 98, 86, 163, 74, 180, 173])

# ============================================
# Seeds
# ============================================

POOL_SEED = b"pool"
POOL_VAULT_SEED = b"pool_vault"
OBSERVATION_STATE_SEED = b"observation"

# ============================================
# PDA Derivation Functions
# ============================================

def get_pool_pda(amm_config: Pubkey, mint1: Pubkey, mint2: Pubkey) -> Pubkey:
    """
    Derive the pool PDA for a given amm_config and two mints.
    Seeds: ["pool", amm_config, mint1, mint2]
    """
    seeds = [POOL_SEED, bytes(amm_config), bytes(mint1), bytes(mint2)]
    (pda, _) = Pubkey.find_program_address(seeds, RAYDIUM_CPMM_PROGRAM_ID)
    return pda


def get_vault_pda(pool_state: Pubkey, mint: Pubkey) -> Pubkey:
    """
    Derive the vault PDA for a given pool and mint.
    Seeds: ["pool_vault", pool_state, mint]
    """
    seeds = [POOL_VAULT_SEED, bytes(pool_state), bytes(mint)]
    (pda, _) = Pubkey.find_program_address(seeds, RAYDIUM_CPMM_PROGRAM_ID)
    return pda


def get_observation_state_pda(pool_state: Pubkey) -> Pubkey:
    """
    Derive the observation state PDA for a given pool.
    Seeds: ["observation", pool_state]
    """
    seeds = [OBSERVATION_STATE_SEED, bytes(pool_state)]
    (pda, _) = Pubkey.find_program_address(seeds, RAYDIUM_CPMM_PROGRAM_ID)
    return pda


# ============================================
# Raydium CPMM Parameters Dataclass
# ============================================

@dataclass
class RaydiumCpmmParams:
    """Parameters for Raydium CPMM protocol trading."""
    pool_state: Optional[Pubkey] = None  # If None, will derive from mints and amm_config
    amm_config: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    base_mint: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    quote_mint: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    base_reserve: int = 0
    quote_reserve: int = 0
    base_vault: Optional[Pubkey] = None
    quote_vault: Optional[Pubkey] = None
    base_token_program: Pubkey = TOKEN_PROGRAM
    quote_token_program: Pubkey = TOKEN_PROGRAM
    observation_state: Optional[Pubkey] = None

    @property
    def is_wsol(self) -> bool:
        """Check if the pool contains WSOL."""
        return self.base_mint == WSOL_TOKEN_ACCOUNT or self.quote_mint == WSOL_TOKEN_ACCOUNT

    @property
    def is_usdc(self) -> bool:
        """Check if the pool contains USDC."""
        return self.base_mint == USDC_TOKEN_ACCOUNT or self.quote_mint == USDC_TOKEN_ACCOUNT


# ============================================
# Raydium CPMM Calculation Functions
# ============================================

def compute_swap_amount(
    base_reserve: int,
    quote_reserve: int,
    is_base_in: bool,
    amount_in: int,
    slippage_bps: int,
) -> tuple:
    """
    Compute swap output amount for Raydium CPMM.
    Returns (amount_out, min_amount_out).
    """
    # Apply trade fee (0.25%)
    fee_rate = TRADE_FEE_RATE / FEE_RATE_DENOMINATOR_VALUE
    amount_in_after_fee = int(amount_in * (1 - fee_rate))

    if is_base_in:
        # Base -> Quote
        # out = (amount_in * quote_reserve) / (base_reserve + amount_in)
        numerator = amount_in_after_fee * quote_reserve
        denominator = base_reserve + amount_in_after_fee
        amount_out = numerator // denominator
    else:
        # Quote -> Base
        # out = (amount_in * base_reserve) / (quote_reserve + amount_in)
        numerator = amount_in_after_fee * base_reserve
        denominator = quote_reserve + amount_in_after_fee
        amount_out = numerator // denominator

    # Apply slippage
    min_amount_out = calculate_with_slippage_sell(amount_out, slippage_bps)

    return (amount_out, min_amount_out)


# ============================================
# Build Buy Instructions
# ============================================

def build_buy_instructions(
    payer: Pubkey,
    output_mint: Pubkey,
    input_amount: int,
    params: RaydiumCpmmParams,
    slippage_bps: int = DEFAULT_SLIPPAGE,
    create_input_ata: bool = True,
    create_output_ata: bool = True,
    close_input_ata: bool = False,
    fixed_output_amount: Optional[int] = None,
) -> List[Instruction]:
    """
    Build Raydium CPMM buy instructions.

    Args:
        payer: The wallet paying for the swap
        output_mint: The token mint to buy
        input_amount: Amount of SOL/USDC to spend
        params: Raydium CPMM protocol parameters
        slippage_bps: Slippage tolerance in basis points
        create_input_ata: Whether to create WSOL ATA if needed
        create_output_ata: Whether to create output token ATA if needed
        close_input_ata: Whether to close WSOL ATA after swap
        fixed_output_amount: If set, use this as exact output amount

    Returns:
        List of instructions for the buy operation
    """
    if input_amount == 0:
        raise ValueError("Amount cannot be zero")

    instructions = []

    # Validate pool contains WSOL or USDC
    if not params.is_wsol and not params.is_usdc:
        raise ValueError("Pool must contain WSOL or USDC")

    # Get pool state
    pool_state = params.pool_state
    if pool_state is None:
        pool_state = get_pool_pda(params.amm_config, params.base_mint, params.quote_mint)

    # Determine if base is input (WSOL/USDC)
    is_base_in = params.base_mint == WSOL_TOKEN_ACCOUNT or params.base_mint == USDC_TOKEN_ACCOUNT

    # Get output token program
    mint_token_program = params.quote_token_program if is_base_in else params.base_token_program

    # Calculate swap amount
    _, min_amount_out = compute_swap_amount(
        params.base_reserve,
        params.quote_reserve,
        is_base_in,
        input_amount,
        slippage_bps,
    )

    if fixed_output_amount is not None:
        minimum_amount_out = fixed_output_amount
    else:
        minimum_amount_out = min_amount_out

    # Determine input mint (WSOL or USDC)
    input_mint = WSOL_TOKEN_ACCOUNT if params.is_wsol else USDC_TOKEN_ACCOUNT

    # Get user token accounts
    input_token_account = get_associated_token_address(payer, input_mint, TOKEN_PROGRAM)
    output_token_account = get_associated_token_address(payer, output_mint, mint_token_program)

    # Get vaults
    input_vault = params.base_vault if is_base_in else params.quote_vault
    if input_vault is None:
        input_vault = get_vault_pda(pool_state, input_mint)

    output_vault = params.quote_vault if is_base_in else params.base_vault
    if output_vault is None:
        output_vault = get_vault_pda(pool_state, output_mint)

    # Get observation state
    observation_state = params.observation_state
    if observation_state is None:
        observation_state = get_observation_state_pda(pool_state)

    # Handle WSOL if needed
    if create_input_ata and params.is_wsol:
        instructions.extend(handle_wsol(payer, input_amount))

    # Create output ATA if needed
    if create_output_ata:
        instructions.append(
            create_associated_token_account_idempotent_instruction(
                payer, payer, output_mint, mint_token_program
            )
        )

    # Build instruction data
    data = SWAP_BASE_IN_DISCRIMINATOR + struct.pack("<QQ", input_amount, minimum_amount_out)

    # Build accounts list
    accounts = [
        AccountMeta(payer, True, True),  # payer (signer, writable)
        AccountMeta(AUTHORITY, False, False),  # authority (readonly)
        AccountMeta(params.amm_config, False, False),  # amm_config (readonly)
        AccountMeta(pool_state, False, True),  # pool_state (writable)
        AccountMeta(input_token_account, False, True),  # input_token_account (writable)
        AccountMeta(output_token_account, False, True),  # output_token_account (writable)
        AccountMeta(input_vault, False, True),  # input_vault (writable)
        AccountMeta(output_vault, False, True),  # output_vault (writable)
        AccountMeta(TOKEN_PROGRAM, False, False),  # input_token_program (readonly)
        AccountMeta(mint_token_program, False, False),  # output_token_program (readonly)
        AccountMeta(input_mint, False, False),  # input_token_mint (readonly)
        AccountMeta(output_mint, False, False),  # output_token_mint (readonly)
        AccountMeta(observation_state, False, True),  # observation_state (writable)
    ]

    instructions.append(Instruction(RAYDIUM_CPMM_PROGRAM_ID, data, accounts))

    # Close WSOL ATA if requested
    if close_input_ata and params.is_wsol:
        instructions.extend(close_wsol(payer))

    return instructions


# ============================================
# Build Sell Instructions
# ============================================

def build_sell_instructions(
    payer: Pubkey,
    input_mint: Pubkey,
    input_amount: int,
    params: RaydiumCpmmParams,
    slippage_bps: int = DEFAULT_SLIPPAGE,
    create_output_ata: bool = True,
    close_output_ata: bool = False,
    close_input_ata: bool = False,
    fixed_output_amount: Optional[int] = None,
) -> List[Instruction]:
    """
    Build Raydium CPMM sell instructions.

    Args:
        payer: The wallet paying for the swap
        input_mint: The token mint to sell
        input_amount: Amount of tokens to sell
        params: Raydium CPMM protocol parameters
        slippage_bps: Slippage tolerance in basis points
        create_output_ata: Whether to create WSOL ATA for receiving SOL
        close_output_ata: Whether to close WSOL ATA after swap
        close_input_ata: Whether to close token ATA after swap
        fixed_output_amount: If set, use this as exact output amount

    Returns:
        List of instructions for the sell operation
    """
    if input_amount == 0:
        raise ValueError("Amount cannot be zero")

    instructions = []

    # Validate pool contains WSOL or USDC
    if not params.is_wsol and not params.is_usdc:
        raise ValueError("Pool must contain WSOL or USDC")

    # Get pool state
    pool_state = params.pool_state
    if pool_state is None:
        pool_state = get_pool_pda(params.amm_config, params.base_mint, params.quote_mint)

    # Determine if quote is output (WSOL/USDC)
    is_quote_out = params.quote_mint == WSOL_TOKEN_ACCOUNT or params.quote_mint == USDC_TOKEN_ACCOUNT

    # Get input token program
    mint_token_program = params.base_token_program if is_quote_out else params.quote_token_program

    # Calculate swap amount
    _, min_amount_out = compute_swap_amount(
        params.base_reserve,
        params.quote_reserve,
        is_quote_out,  # Swap direction is reversed for sell
        input_amount,
        slippage_bps,
    )

    if fixed_output_amount is not None:
        minimum_amount_out = fixed_output_amount
    else:
        minimum_amount_out = min_amount_out

    # Determine output mint (WSOL or USDC)
    output_mint = WSOL_TOKEN_ACCOUNT if params.is_wsol else USDC_TOKEN_ACCOUNT

    # Get user token accounts
    output_token_account = get_associated_token_address(payer, output_mint, TOKEN_PROGRAM)
    input_token_account = get_associated_token_address(payer, input_mint, mint_token_program)

    # Get vaults
    output_vault = params.quote_vault if is_quote_out else params.base_vault
    if output_vault is None:
        output_vault = get_vault_pda(pool_state, output_mint)

    input_vault = params.base_vault if is_quote_out else params.quote_vault
    if input_vault is None:
        input_vault = get_vault_pda(pool_state, input_mint)

    # Get observation state
    observation_state = params.observation_state
    if observation_state is None:
        observation_state = get_observation_state_pda(pool_state)

    # Create WSOL ATA if needed for receiving SOL
    if create_output_ata and params.is_wsol:
        instructions.append(
            create_associated_token_account_idempotent_instruction(
                payer, payer, WSOL_TOKEN_ACCOUNT, TOKEN_PROGRAM
            )
        )

    # Build instruction data
    data = SWAP_BASE_IN_DISCRIMINATOR + struct.pack("<QQ", input_amount, minimum_amount_out)

    # Build accounts list
    accounts = [
        AccountMeta(payer, True, True),  # payer (signer, writable)
        AccountMeta(AUTHORITY, False, False),  # authority (readonly)
        AccountMeta(params.amm_config, False, False),  # amm_config (readonly)
        AccountMeta(pool_state, False, True),  # pool_state (writable)
        AccountMeta(input_token_account, False, True),  # input_token_account (writable)
        AccountMeta(output_token_account, False, True),  # output_token_account (writable)
        AccountMeta(input_vault, False, True),  # input_vault (writable)
        AccountMeta(output_vault, False, True),  # output_vault (writable)
        AccountMeta(mint_token_program, False, False),  # input_token_program (readonly)
        AccountMeta(TOKEN_PROGRAM, False, False),  # output_token_program (readonly)
        AccountMeta(input_mint, False, False),  # input_token_mint (readonly)
        AccountMeta(output_mint, False, False),  # output_token_mint (readonly)
        AccountMeta(observation_state, False, True),  # observation_state (writable)
    ]

    instructions.append(Instruction(RAYDIUM_CPMM_PROGRAM_ID, data, accounts))

    # Close WSOL ATA if requested
    if close_output_ata and params.is_wsol:
        instructions.extend(close_wsol(payer))

    # Close token ATA if requested
    if close_input_ata:
        instructions.append(
            close_token_account_instruction(
                mint_token_program,
                input_token_account,
                payer,
                payer,
            )
        )

    return instructions


# ============================================
# Exports
# ============================================

__all__ = [
    # Program IDs and Constants
    "RAYDIUM_CPMM_PROGRAM_ID",
    "AUTHORITY",
    "FEE_RATE_DENOMINATOR_VALUE",
    "TRADE_FEE_RATE",
    "CREATOR_FEE_RATE",
    "PROTOCOL_FEE_RATE",
    "FUND_FEE_RATE",
    # Discriminators
    "SWAP_BASE_IN_DISCRIMINATOR",
    "SWAP_BASE_OUT_DISCRIMINATOR",
    # PDA Functions
    "get_pool_pda",
    "get_vault_pda",
    "get_observation_state_pda",
    # Params
    "RaydiumCpmmParams",
    # Calculation Functions
    "compute_swap_amount",
    # Instruction Builders
    "build_buy_instructions",
    "build_sell_instructions",
]
